stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/options/message/MessageOptionItemsFactory.kt

```diff
@@ -0,0 +1,202 @@
+import io.getstream.chat.android.client.models.Message
+import io.getstream.chat.android.client.models.User
+import io.getstream.chat.android.client.utils.SyncStatus
+import io.getstream.chat.android.common.state.Copy
+import io.getstream.chat.android.common.state.Delete
+import io.getstream.chat.android.common.state.Edit
+import io.getstream.chat.android.common.state.Flag
+import io.getstream.chat.android.common.state.MuteUser
+import io.getstream.chat.android.common.state.Pin
+import io.getstream.chat.android.common.state.Reply
+import io.getstream.chat.android.common.state.Resend
+import io.getstream.chat.android.common.state.ThreadReply
+import io.getstream.chat.android.ui.R
+import io.getstream.chat.android.ui.common.extensions.internal.getDrawableCompat
+import io.getstream.chat.android.ui.common.extensions.internal.hasLink
+import io.getstream.chat.android.ui.message.list.MessageListViewStyle
+
+/**
+ * An interface that allows the creation of message option items.
+ */
+public interface MessageOptionItemsFactory {
+
+    /**
+     * Creates [MessageOptionItem]s for the selected message.
+     *
+     * @param selectedMessage The currently selected message.
+     * @param currentUser The currently logged in user.
+     * @param isInThread If the message is being displayed in a thread.
+     * @param ownCapabilities Set of capabilities the user is given for the current channel.
+     * @param style The style to be applied to the view.
+     * @return The list of message option items to display.
+     */
+    public fun createMessageOptionItems(
+        selectedMessage: Message,
+        currentUser: User?,
+        isInThread: Boolean,
+        ownCapabilities: Set<String>,
+        style: MessageListViewStyle,
+    ): List<MessageOptionItem>
+
+    public companion object {
+        /**
+         * Builds the default message option items factory.
+         *
+         * @return The default implementation of [MessageOptionItemsFactory].
+         */
+        public fun defaultFactory(context: Context): MessageOptionItemsFactory {
+            return DefaultMessageOptionItemsFactory(context)
+        }
+    }
+}
+
+/**
+ * The default implementation of [MessageOptionItemsFactory].
+ *
+ * @param context The context to load resources.
+ */
+public open class DefaultMessageOptionItemsFactory(
+    private val context: Context,
+) : MessageOptionItemsFactory {
+
+    /**
+     * Creates [MessageOptionItem]s for the selected message.
+     *
+     * @param selectedMessage The currently selected message.
+     * @param currentUser The currently logged in user.
+     * @param isInThread If the message is being displayed in a thread.
+     * @param ownCapabilities Set of capabilities the user is given for the current channel.
+     * @param style The style to be applied to the view.
+     * @return The list of message option items to display.
+     */
+    override fun createMessageOptionItems(
```

Comment:
Basically, "hoisting" the logic to `MessageListView` so that the overlay is as dumb as possible.